File: app/src/main/java/com/friendlypos/reenvio_email/fragment/EmailSelecClienteFragment.java
package com.friendlypos.reenvio_email.fragment;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.friendlypos.R;
import com.friendlypos.distribucion.activity.DistribucionActivity;
import com.friendlypos.distribucion.adapters.DistrClientesAdapter;
import com.friendlypos.distribucion.adapters.DistrResumenAdapter;
import com.friendlypos.distribucion.fragment.BaseFragment;
import com.friendlypos.distribucion.fragment.DistSelecProductoFragment;
import com.friendlypos.distribucion.modelo.sale;
import com.friendlypos.principal.modelo.Clientes;
import com.friendlypos.reenvio_email.activity.EmailActivity;
import com.friendlypos.reenvio_email.adapters.EmailClientesAdapter;
import com.friendlypos.reimpresion_pedidos.activity.ReimprimirPedidosActivity;

import java.util.ArrayList;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import io.realm.Realm;
import io.realm.RealmQuery;
import io.realm.RealmResults;

import static io.realm.internal.SyncObjectServerFacade.getApplicationContext;

public class EmailSelecClienteFragment extends BaseFragment implements SearchView.OnQueryTextListener {

    private Realm realm;
    @Bind(R.id.recyclerViewEmailCliente)
    public RecyclerView recyclerView;
    private EmailClientesAdapter adapter;

    @Override
    public void onDestroy() {
        super.onDestroy();
    }
    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.fragment_email_selec_cliente, container,
                false);
        setHasOptionsMenu(true);
        ButterKnife.bind(this, rootView);
        return rootView;
        }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if(adapter == null) {
            adapter = new EmailClientesAdapter(getContext(), ((EmailActivity) getActivity()), getListClientes());
            //adapter2 = new DistrResumenAdapter();
        }
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setAdapter(adapter);
        Log.d("listadeClientes", getListClientes() + "");
    }

    private List<Clientes> getListClientes(){

        realm = Realm.getDefaultInstance();
        RealmQuery<Clientes> query = realm.where(Clientes.class);
        RealmResults<Clientes> result1 = query.findAll();
        if(result1.size() == 0){
            Toast.makeText(getApplicationContext(),"Favor descargar datos primero",Toast.LENGTH_LONG).show();
        }
        return result1;

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
         realm.close();
    }

    @Override
    public void updateData() {
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_email, menu);

        final MenuItem item = menu.findItem(R.id.action_search);
        final SearchView searchView = (SearchView) MenuItemCompat.getActionView(item);
        searchView.setOnQueryTextListener(this);

        MenuItemCompat.setOnActionExpandListener(item,
                new MenuItemCompat.OnActionExpandListener() {
                    @Override
                    public boolean onMenuItemActionCollapse(MenuItem item) {
                        // Do something when collapsed
                        adapter.setFilter(getListClientes());
                        return true; // Return true to collapse action view
                    }

                    @Override
                    public boolean onMenuItemActionExpand(MenuItem item) {
                        // Do something when expanded
                        return true; // Return true to expand action view
                    }
                });
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        final List<Clientes> filteredModelList = filter(getListClientes(), newText);
        adapter.setFilter(filteredModelList);
        return true;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }


    private List<Clientes> filter(List<Clientes> models, String query) {
        query = query.toLowerCase();

        final List<Clientes> filteredModelList = new ArrayList<>();
        for (Clientes model : models) {
            final String text = model.getFantasyName().toLowerCase();
            Log.d("FiltroEmail", text);
            if (text.contains(query)) {
                filteredModelList.add(model);
            }
        }
        return filteredModelList;
    }


}

File: app/src/main/java/com/friendlypos/reenvio_email/fragment/EmailSelecClienteFragment.kt
package com.friendlysystemgroup.friendlypos.reenvio_email.fragment

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuItemCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.friendlysystemgroup.friendlypos.R
import com.friendlysystemgroup.friendlypos.databinding.FragmentEmailSelecClienteBinding
import com.friendlysystemgroup.friendlypos.distribucion.fragment.BaseFragment
import com.friendlysystemgroup.friendlypos.principal.modelo.Clientes
import com.friendlysystemgroup.friendlypos.Reenvio_email.activity.EmailActivity
import com.friendlysystemgroup.friendlypos.Reenvio_email.adapters.EmailClientesAdapter
import io.realm.Realm
import io.realm.RealmResults
import java.util.Locale

class EmailSelecClienteFragment : BaseFragment(), SearchView.OnQueryTextListener {
    private var binding: FragmentEmailSelecClienteBinding? = null
    private var realm: Realm? = null
    private var adapter: EmailClientesAdapter? = null
    private var clientes: List<Clientes> = listOf()
    private val clientesFilter: MutableList<Clientes> = mutableListOf()

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentEmailSelecClienteBinding.inflate(inflater, container, false)
        setHasOptionsMenu(true)
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupRecyclerView()
    }

    private fun setupRecyclerView() {
        binding?.recyclerViewEmailCliente?.apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(activity)
            
            clientes = listClientes
            clientesFilter.clear()
            clientesFilter.addAll(clientes)
            
            adapter = EmailClientesAdapter(
                context, 
                activity as EmailActivity?,
                clientesFilter
            ).also {
                this@EmailSelecClienteFragment.adapter = it
            }
        }
        
        Log.d("listadeClientes", "Clientes cargados: ${clientes.size}")
    }

    private val listClientes: List<Clientes>
        get() {
            val realm = Realm.getDefaultInstance()
            try {
                val result: RealmResults<Clientes> = realm.where(Clientes::class.java).findAll()
                
                if (result.isEmpty()) {
                    Toast.makeText(
                        context,
                        "Favor descargar datos primero",
                        Toast.LENGTH_LONG
                    ).show()
                    return emptyList()
                }
                
                return realm.copyFromRealm(result)
            } finally {
                realm.close()
            }
        }

    override fun onDestroyView() {
        super.onDestroyView()
        realm?.close()
        binding = null
    }

    override fun updateData() {
        // Implementar si es necesario
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_email, menu)

        val item = menu.findItem(R.id.action_search)
        val searchView = MenuItemCompat.getActionView(item) as SearchView
        searchView.setOnQueryTextListener(this)

        MenuItemCompat.setOnActionExpandListener(
            item,
            object : MenuItemCompat.OnActionExpandListener {
                override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                    adapter?.setFilter(listClientes)
                    return true
                }

                override fun onMenuItemActionExpand(item: MenuItem): Boolean {
                    return true
                }
            })
    }

    override fun onQueryTextChange(newText: String): Boolean {
        filterList(newText)
        return true
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        return false
    }

    private fun filterList(query: String) {
        clientesFilter.clear()
        
        if (query.isEmpty()) {
            clientesFilter.addAll(clientes)
        } else {
            val lowerCaseQuery = query.lowercase(Locale.getDefault())
            
            clientes.forEach { cliente ->
                cliente.fantasyName?.let { name ->
                    if (name.lowercase(Locale.getDefault()).contains(lowerCaseQuery)) {
                        clientesFilter.add(cliente)
                    }
                }
            }
        }
        
        adapter?.setFilter(clientesFilter)
    }
}
